libraries/chain/block_log.cpp

```diff
@@ -31,1039 +21,1291 @@ namespace eosio { namespace chain {
+            catalog.open(log_dir, config.retained_dir, config.archive_dir, "blocks");
+            catalog.max_retained_files = config.max_retained_files;
 
-      my->check_open_files();
-
-      uint64_t pos;
-
-      // Check that the file is not empty
-      my->block_file.seek_end(0);
-      if (my->block_file.tellp() <= sizeof(pos))
-         return {};
-
-      //figure out if this is a pruned log or not. we can't just look at the configuration since
-      // read_head() is called early on, and this isn't hot enough to warrant a member bool to track it
-      my->block_file.seek(0);
-      uint32_t current_version;
-      fc::raw::unpack(my->block_file, current_version);
-      const bool is_currently_pruned = detail::is_pruned_log_and_mask_version(current_version);
-
-      my->block_file.seek_end(0);
-      if(is_currently_pruned)
-         my->block_file.skip(-sizeof(uint32_t)); //skip the trailer containing block count
-      my->block_file.skip(-sizeof(pos));
-      fc::raw::unpack(my->block_file, pos);
-      if (pos != npos)
-         return read_block(pos);
-      return {};
-   }
+            open(log_dir, config.fix_irreversible_blocks);
+            const auto log_size = fc::file_size(block_file.get_file_path());
 
-   const signed_block_ptr& block_log::head()const {
-      return my->head;
-   }
+            if (log_size == 0 && !catalog.empty()) {
+               basic_block_log::reset(catalog.verifier.chain_id, catalog.last_block_num() + 1);
+               update_head(read_block_by_num(catalog.last_block_num()));
+            } else {
+               EOS_ASSERT(catalog.verifier.chain_id.empty() || catalog.verifier.chain_id == preamble.chain_id(),
+                          block_log_exception, "block log file ${path} has a different chain id",
+                          ("path", block_file.get_file_path()));
+            }
+         }
 
-   const block_id_type&    block_log::head_id()const {
-      return my->head_id;
-   }
+         void split_log() {
+            block_file.close();
+            index_file.close();
 
-   uint32_t block_log::first_block_num() const {
-      return my->first_block_num;
-   }
+            catalog.add(preamble.first_block_num, this->head->block_num(), block_file.get_file_path().parent_path(),
+                        "blocks");
 
-   void block_log::construct_index() {
-      if (my->not_generate_block_log) {
-         ilog("Not need to construct index in no blocks.log mode (block-log-retain-blocks=0)");
-         return;
-      }
+            block_file.open(fc::cfile::truncate_rw_mode);
+            index_file.open(fc::cfile::truncate_rw_mode);
+            preamble.ver             = block_log::max_supported_version;
+            preamble.chain_context   = preamble.chain_id();
+            preamble.first_block_num = this->head->block_num() + 1;
+            preamble.write_to(block_file);
+            flush();
+         }
```

Comment:
Can `split_log()` be made safer so that it first tries to open a new file, and if that fails, continues logging to the current file?

Reply:
Agreed. Had similar comments to #628.